### PR TITLE
video images cleanup

### PR DESCRIPTION
**Pull request by muhammad-ammar**
[EDUCATOR-212](https://openedx.atlassian.net/browse/EDUCATOR-212)

### CODE REVIEW

**Comment by muhammad-ammar** [comment]
@muzaffaryousaf @mushtaqak @Qubad786 This is ready for review.

File: edxval/models.py [hunk 1]
```diff
@@ -286,6 +286,13 @@ def create_or_update(cls, course_video, file_name, image_data=None):
         """
         video_image, created = cls.objects.get_or_create(course_video=course_video)
         if image_data:
+            # Delete the existing image only if this image is not used by anyone else. This is necessary because
+            # after a course re-run, a video in original course and the new course points to same image, So when
+            # we update an image in new course and delete the existing image. This will delete the image from
+            # original course also leaving video with no image.
```
Comment (mushtaqak) [inline edxval/models.py]: last line can be made a bit better : `original course as well, thus leaving video with having no image`

File: edxval/tests/test_api.py [hunk 2]
```diff
@@ -1386,3 +1386,30 @@ def test_video_image_urls_field_validation(self):
...
+        """
+        # Set and verify both course video images to have same image
+        shared_image = self.course_video.video_image.image.name = self.course_video2.video_image.image.name
+        self.course_video.video_image.save()
+        self.assertEqual(self.course_video.video_image.image.name, self.course_video2.video_image.image.name)
+
+        api.update_video_image(self.edx_video_id, self.course_id, ImageFile(open(self.image_path2)), 'image.jpg')
+
+        # Verify share image is not deleted by opening the file
+        ImageFile(open(shared_image))
```
Comment (mushtaqak) [inline edxval/tests/test_api.py]: I think we should also check old course has old image but new course has the latest image now.

File: edxval/tests/test_api.py [hunk 3]
```diff
@@ -1386,3 +1386,30 @@ def test_video_image_urls_field_validation(self):
...
+
+    def test_video_image_deletion_single(self):
+        """
+        Test video image deletion works as expected when an image is used by only the video being updated.
+        """
+        existing_image_name = self.course_video.video_image.image.name
+
+        api.update_video_image(self.edx_video_id, self.course_id, ImageFile(open(self.image_path2)), 'image.jpg')
+
+        with self.assertRaises(IOError) as file_open_exception:
```
Comment (mushtaqak) [inline edxval/tests/test_api.py]: I would add a comment that this is need to test that image has been deleted from storage.

File: edxval/tests/test_api.py [hunk 4]
```diff
@@ -1386,3 +1386,30 @@ def test_video_image_urls_field_validation(self):
...
+            ImageFile(open(existing_image_name))
+
+        self.assertEqual(file_open_exception.exception.strerror, u'No such file or directory')
```
Comment (mushtaqak) [inline edxval/tests/test_api.py]: I think we also should check video image is updated.

File: edxval/tests/test_api.py [hunk 5]
```diff
@@ -1386,3 +1386,30 @@ def test_video_image_urls_field_validation(self):
...
+
+    def test_video_image_deletion_multiple(self):
+        """
+        Test video image deletion works as expected if an image is used by multiple course videos.
+        """
+        # Set and verify both course video images to have same image
+        shared_image = self.course_video.video_image.image.name = self.course_video2.video_image.image.name
+        self.course_video.video_image.save()
```
Comment (mushtaqak) [inline edxval/tests/test_api.py]: In above test we are using `api.update_video_image` method to save, is there any particular reasons ?

File: edxval/tests/test_api.py (hunk 2 above)
Comment (mushtaqak) [inline edxval/tests/test_api.py]: Also, we can get image file object and check it's None ? I don't know if this is the best way to test file exists?

**Comment by muhammad-ammar** [comment]
@mushtaqak addressed your feedback.

File: edxval/tests/test_api.py [hunk 6]
```diff
@@ -1386,3 +1386,47 @@ def test_video_image_urls_field_validation(self):
...
+        self.assertEqual(self.course_video.video_image.image.name, self.course_video2.video_image.image.name)
+
+        # This will replace the image for self.course_video but image will
+        # not be deleted becuase it is also used by self.course_video2
```
Comment (Qubad786) [inline edxval/tests/test_api.py]: typo: because

**Comment by mushtaqak** [comment]
Looks good 👍

**Pull request merged**